sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldContainAssertion.cs

```diff
@@ -0,0 +1,128 @@
+
+        private static readonly ISet<KnownType> TrackedIgnoreAttributes = ImmutableHashSet.Create(
+            KnownType.Microsoft_VisualStudio_TestTools_UnitTesting_IgnoreAttribute,
+            KnownType.NUnit_Framework_IgnoreAttribute);
+
+        private static readonly IEnumerable<string> TrackedAssertionMethodName = ImmutableList.Create(
+            "Assert", "Should", "Expect", "Must", "Verify", "Validate");
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var methodDeclaration = (MethodDeclarationSyntax)c.Node;
+                    if (methodDeclaration.Identifier.IsMissing ||
+                        methodDeclaration.Body == null ||
+                        !c.IsTest())
+                    {
+                        return;
+                    }
+
+                    var methodSymbol = c.SemanticModel.GetDeclaredSymbol(methodDeclaration);
+                    var attributes = methodSymbol?.GetAttributes();
+                    if (methodSymbol == null ||
+                        !IsTestMethod(attributes) ||
+                        HasExpectedExceptionAttribute(attributes) ||
+                        IsTestIgnored(attributes))
+                    {
+                        return;
+                    }
+
+                    var hasAnyAssertion = methodDeclaration.DescendantNodes()
+                        .OfType<InvocationExpressionSyntax>()
+                        .Any(invocation => IsAssertion(invocation, c.SemanticModel));
+                    if (!hasAnyAssertion)
+                    {
+                        c.ReportDiagnostic(Diagnostic.Create(rule, methodDeclaration.Identifier.GetLocation()));
+                    }
+                }, SyntaxKind.MethodDeclaration);
+        }
+
+        private static bool IsTestMethod(IEnumerable<AttributeData> attributes) =>
+            attributes.Any(a => a.AttributeClass.IsAny(TrackedTestAttributes));
+
+        private static bool HasExpectedExceptionAttribute(IEnumerable<AttributeData> attributes) =>
+            attributes.Any(a => a.AttributeClass.IsAny(ExpectedExceptionAttributes));
+
+        private static bool IsTestIgnored(IEnumerable<AttributeData> attributes)
+        {
+            if (attributes.Any(a => a.AttributeClass.IsAny(TrackedIgnoreAttributes)))
+            {
+                return true;
+            }
+
+            var factAttributeSyntax = attributes.FirstOrDefault(a => a.AttributeClass.Is(KnownType.Xunit_FactAttribute))
+                ?.ApplicationSyntaxReference.GetSyntax() as AttributeSyntax;
+
+            return factAttributeSyntax?.ArgumentList != null &&
+                factAttributeSyntax.ArgumentList.Arguments.Any(x => x.NameEquals.Name.Identifier.ValueText == "Skip");
+        }
+
+        private static bool IsAssertion(InvocationExpressionSyntax invocation, SemanticModel model)
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Consider using more general type 'Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax' instead of 'Microsoft.CodeAnalysis.CSharp.Syntax.InvocationExpressionSyntax'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS3242)